### PR TITLE
Run tests on py3

### PR DESCRIPTION
**Pull request by KyleJu**
- Separate py2 (sendemail_test.py) and py3 tests 
- Get tests to run on py3
- Set up CI for py2 test target and py3 test target
- additional test coverage to catch py3 issues??

### CODE REVIEW

File: testing_config.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 app_engine_path = os.environ.get('APP_ENGINE_PATH', '')
 if not app_engine_path:
-  app_engine_path = '/usr/lib/google-cloud-sdk/platform/google_appengine'
+  app_engine_path = '/usr/local/google/home/kyleju/google-cloud-sdk/platform/google_appengine'
```
Comment (KyleJu) [inline testing_config.py:25]: revert this after done

**Comment by KyleJu** [comment]
Down to 14 errors and 3 failing tests.

File: internals/models.py [hunk 2]
```diff
@@ -349,7 +347,7 @@ def del_none(d):
   """
   Delete dict keys with None values, and empty lists, recursively.
   """
-  for key, value in d.items():
+  for key, value in list(d.items()):
```
Comment (jrobbins) [inline internals/models.py:350]: I thought that we could use items() in python3.  E.g., from the python3 tutorial:
https://docs.python.org/3/tutorial/datastructures.html#looping-techniques
And, when I try it in python3 interactively, items() seems to work.

Comment (jrobbins) [reply internals/models.py:350]: I think you need the list() only when the list is being stored in a variable, not when it is used only for iteration in a for-loop.

**Comment by KyleJu** [comment]
CI made some progress:

EEEEEEEEEE............EETraceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/main.py", line 271, in runTests
    self.result = testRunner.run(self.test)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/runner.py", line 176, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/case.py", line 651, in __call__
    return self.run(*args, **kwds)
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/testing_config.py", line 108, in run
    client = ndb.Client()
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 107, in __init__
    super(Client, self).__init__(
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/client.py", line 318, in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/client.py", line 266, in __init__
    project = self._determine_default(project)
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 212, in _determine_default
    return _determine_default_project(project)
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 67, in _determine_default_project
    project = _helpers._determine_default_project(project=project)
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/_helpers.py", line 186, in _determine_default_project
    _, project = google.auth.default()
  File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/auth/_default.py", line 356, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

**Comment by KyleJu** [comment]
> CI made some progress:
> 
> EEEEEEEEEE............EETraceback (most recent call last):
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/runpy.py", line 197, in _run_module_as_main
> return _run_code(code, main_globals, None,
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/runpy.py", line 87, in _run_code
> exec(code, run_globals)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/**main**.py", line 18, in
> main(module=None)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/main.py", line 101, in **init**
> self.runTests()
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/main.py", line 271, in runTests
> self.result = testRunner.run(self.test)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/runner.py", line 176, in run
> test(result)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in **call**
> return self.run(*args, **kwds)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
> test(result)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in **call**
> return self.run(*args, **kwds)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
> test(result)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 84, in **call**
> return self.run(*args, **kwds)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/suite.py", line 122, in run
> test(result)
> File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/unittest/case.py", line 651, in **call**
> return self.run(*args, **kwds)
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/testing_config.py", line 108, in run
> client = ndb.Client()
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 107, in **init**
> super(Client, self).**init**(
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/client.py", line 318, in **init**
> _ClientProjectMixin.**init**(self, project=project, credentials=credentials)
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/client.py", line 266, in **init**
> project = self._determine_default(project)
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 212, in _determine_default
> return _determine_default_project(project)
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/ndb/client.py", line 67, in _determine_default_project
> project = _helpers._determine_default_project(project=project)
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/cloud/_helpers.py", line 186, in _determine_default_project
> _, project = google.auth.default()
> File "/home/runner/work/chromium-dashboard/chromium-dashboard/lib/google/auth/_default.py", line 356, in default
> raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
> google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

Looks like it is related to [ndb setup in py3](https://googleapis.dev/python/python-ndb/0.2.0/index.html)

File: testing_config.py [hunk 3]
```diff
@@ -25,7 +25,7 @@
   app_engine_path = '/usr/lib/google-cloud-sdk/platform/google_appengine'
 if not os.path.exists(app_engine_path):
```
Comment (KyleJu) [inline testing_config.py:26]: Remove this `if` after done

File: internals/models.py [hunk 4]
```diff
@@ -924,8 +922,8 @@ def getSortingMilestone(feature):
       shipping_features = map(getSortingMilestone, shipping_features)
 
       # First sort by name, then sort by feature milestone (latest first).
-      shipping_features.sort(key=lambda f: f.name, reverse=False)
-      shipping_features.sort(key=lambda f: f._sort_by_milestone, reverse=True)
+      sorted(shipping_features, key=lambda f: f.name, reverse=False)
```
Comment (jrobbins) [inline internals/models.py:925]: The old .sort() method modified the array in-place.  These lines just return a sorted copy without changing the original, so they are no-ops unless we store the result back into the same variable.

Does python 3 not have .sort()?  I thought that it did.

**Comment by KyleJu** [comment]
There are 8 test exceptions and 5 test failures as of 2023ab3. The errors are [here](https://paste.googleplex.com/6081763286712320).

File: internals/models.py (hunk 2 above)
Comment (KyleJu) [reply internals/models.py:350]: Just remembered that it is caused by a different issue

  File "/usr/local/google/home/kyleju/chromium-dashboard/internals/models.py", line 350, in del_none
    for key, value in d.items():
RuntimeError: dictionary changed size during iteration

the d[key]` caused a self-modifying loop. We could use list here or iterate through the keys of the dict

File: internals/models.py (hunk 4 above)
Comment (KyleJu) [reply internals/models.py:925]: shipping_features.sort(key=lambda f: f.name, reverse=False)
AttributeError: 'map' object has no attribute 'sort'

In python 3 map doesn't return a list but an iterator

File: internals/models.py (hunk 2 above)
Comment (jrobbins) [reply internals/models.py:350]: Ah, OK.

File: internals/models.py (hunk 4 above)
Comment (KyleJu) [reply internals/models.py:925]: sorry I didn't address a part of your comment. I should reassign the result of `sorted`

Comment (jrobbins) [reply internals/models.py:925]: Ah, so up on line 922 it should be:
shipping_features = list(map(getSortingMilestone, shipping_features))

Then these two lines can go back to using .sort()

Comment (KyleJu) [reply internals/models.py]: Don't we want to maintain `shipping_features` as a map?

Comment (KyleJu) [reply internals/models.py]: Actually [sorted](https://docs.python.org/3/howto/sorting.html) returns a list so it is the same

**Comment by KyleJu** [comment]
@jrobbins Ready to submit

**Comment by KyleJu** [comment]
Note that `09bed7a` only has 3 test failures. `sorted` changes before caused test failures there

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -18,7 +18,7 @@ jobs:
       - name: set up Python
         uses: actions/setup-python@v2
         with:
-          python-version: '2.7'
+          python-version: '3.x'
```
Comment (KyleJu) [inline .github/workflows/ci.yml:21]: Pin this to 3.9

**Review by jrobbins — APPROVED** [review]

**Pull request merged**